Level-based L0->L0 compaction

ajkr opened the pull request:
Level-based L0->L0 compaction operates on spans of files that aren't currently being compacted. It reduces the number of L0 files, thus making write stall conditions harder to reach.

- L0->L0 is triggered when base level is unavailable due to pending compactions
- L0->L0 always outputs one file of at most `max_level0_burst_file_size` bytes.
- Subcompactions are disabled for L0->L0 since we want to output one file.
- Input files are chosen as the longest span of available files that will fit within the size limit. This minimizes number of files in L0.

Test Plan:

- Database size: 7.4GB
- 8-10GB of page cache is available for benchmark
- Benchmark command: `./db_bench -benchmarks=fillrandom -options_file=/home/andrewkr/myrocks-options -statistics -num=200000000`
- Runs on SSD

before this diff:
```
fillrandom   :       5.758 micros/op 173682 ops/sec;   19.2 MB/s
rocksdb.stall.micros COUNT : 119209777
rocksdb.db.write.micros statistics Percentiles :=> 50 : 3.738630 95 : 6.765740 99 : 12.160575 100 : 2196518.000000
rocksdb.db.write.stall statistics Percentiles :=> 50 : 0.527756 95 : 1008.675013 99 : 1161.926739 100 : 4515.000000
```
after this diff:
```
fillrandom   :       5.261 micros/op 190072 ops/sec;   21.0 MB/s
rocksdb.stall.micros COUNT : 3092710
rocksdb.db.write.micros statistics Percentiles :=> 50 : 3.828868 95 : 6.756103 99 : 11.642088 100 : 380501.000000
rocksdb.db.write.stall statistics Percentiles :=> 50 : 0.527348 95 : 1006.986584 99 : 1161.727554 100 : 2083.000000
```

ajkr commented:
I addressed the changes we talked about. Regarding level0_compactions_in_progress_, this should now contain all compactions from L0, both L0->L0 and L0->base_level. Once we have the other change we discussed -- making L0->base_level choose inputs starting from the oldest L0 file -- we can make this variable specific to L0->base_level compactions. I think we should change the name after that point?

siying commented:
@ajkr sure we can change the name later.

siying commented on db/compaction_picker.cc:
```diff
@@ -1275,6 +1294,41 @@ bool LevelCompactionPicker::PickCompactionBySize(VersionStorageInfo* vstorage,
+    compact_bytes_per_del_file = new_compact_bytes_per_del_file;
+  }
+
+  if (span_len >= kMinFilesForIntraL0Compaction) {
```
Do we also want to introduce a minimum new_compact_bytes_per_del_file or something similar to avoid compacting big files (the files already compacted to be large through L0->L0)?

For example, we can say if new_compact_bytes_per_del_file is larger than write_buffer_size, or file_size_base_target, or level_size_base_target then we are going to skip it.

ajkr replied:
Yes, that's a good idea and might help with the problem I just mentioned. Thanks!

ajkr replied:
Actually we should be ok here. The other change we discussed (taking into account L0 size in compaction score) also prevents the scenario you described.

siying replied:
@ajkr can you explain how?
For example, if every flushed file is 100,
we start by compacting
100 100 100 100
to
400
and then once we have
100 100 100 100 400
we compact to
400 400.
And then we have
400 400 400 400
What prevents us from compacting it to a giant 1600 file?

ajkr replied:
Nothing explicitly prevents. But when L0 has higher compaction score than base_level, no further base_level compactions will be scheduled. In my experiments, compaction-from-base_level was always the cause of L0 being blocked.

ajkr replied:
I realized, though, that my last commit's change in L0 scoring (which I did as a lazy way to fix unit tests) may ruin this. Let me look again.

siying replied:
@ajkr the score for L0 is for L0->L1 compaction. Should we still schedule L0->L0 compaction even if L1->L2 takes priority against L0->L1? I think we can follow it up in the future though. Don't have to do it in the current pull request.

ajkr replied:
I think if L1->L2 takes priority, it'll almost always be successfully scheduled since L1->L2 compactions can run in parallel, thus giving no opportunity for a L0->L0 to run.

ajkr replied:
BTW I fixed the tests, it should be ready for review/commit now.

siying replied:
@ajkr L0->L1 can run in parallel too if subcompaction is enabled. Can you test this case to make sure we don't have regression?

ajkr replied:
Yes. For this case we need to enable both parallel background compactions and subcompactions. Is this a normal config?

If only subcompactions > 1, L0->L0 will never run because there are never running compactions when we schedule a new one.

ajkr replied:
I checked with subcompactions=4 and max_background_compactions=4, the fillrandom throughput still increased 20.1->21.4MB/s.

Also this diff affects compaction pressure token as it uses L0 file count. I'll experiment with making it consider L0 size and see if it helps benchmarks.

siying commented on db/compaction_picker.cc:
```diff
@@ -1275,6 +1294,41 @@ bool LevelCompactionPicker::PickCompactionBySize(VersionStorageInfo* vstorage,
+  inputs->clear();
+  const std::vector<FileMetaData*>& level_files =
+      vstorage->LevelFiles(0 /* level */);
+  if (level_files.size() == 0 || level_files[0]->being_compacted) {
```
I'm thinking about how we can avoid to trigger the test too easily.
For example, we just have 4 files while there happens to be a L1->L2 compaction going on, we don't want compact these 4 files immediately.

Can we raise the standard slightly. Just random thought. How about only triggering it if number of L0 files is no fewer than trigger+2, or trigger * 1.5?

ajkr replied:
Let me try it. The latest benchmark showed a new problem: after compacting large L0 files to base level, the base level's compaction score can become very high (~4). Then L0 compaction won't be scheduled for a long time since its compaction score can't really grow beyond 2.5 (write delay trigger / compaction trigger, so 10 / 4 for MyRocks). Couple ideas:

(1) we can change compaction score to grow exponentially higher in L0 when approaching write delay trigger
(2) we can cap the L0 file size like my initial implementation did. But it's hard to pick a good size and we don't want to add options.

siying replied:
How about we take whichever is larger:
1. current L0 score
2. current L0 size / base_level_size_base.

Because we know have L0->L0 compaction, the size indicated by 1 is now exploded. I think the number can be balanced with 2.

ajkr replied:
This is working well. Thanks 👍

siying commented:
Also please mention it in HISTORY.md.

ajkr commented:
addressed comments and updated benchmark. fyi the cumulative write stall was lower before imposing the restriction of `mutable_cf_options.level0_file_num_compaction_trigger + 2` L0 files to trigger L0->L0 compaction.

ajkr commented:
Any hope to get this in for 5.4?

ajkr commented:
the known drawbacks are (1) slightly increased write-amp in case of heavy writes (~0.5 higher), (2) interaction with dynamic compaction can be a little confusing when individual L0 files trigger compaction to base level (because base level size is between 10% and 100% of configured max base level size, so one L0 file may be bigger than base level's size).
